Sources/NIO/BaseSocketChannel.swift

```diff
@@ -798,7 +798,27 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
     // the `reset` event.
     final func reset() {
         self.readEOF()
-        self.close0(error: ChannelError.eof, mode: .all, promise: nil)
+
+        if self.socket.isOpen {
+            assert(self.lifecycleManager.isRegistered)
+            let error: IOError
+            // if the socket is still registered (and therefore open), let's try to get the actual socket error from the socket
+            do {
+                let result: Int32 = try self.socket.getOption(level: SOL_SOCKET, name: SO_ERROR)
+                if result != 0 {
+                    // we have a socket error, let's forward
+                    // this path will be executed on Linux (EPOLLERR) & Darwin (ev.fflags != 0)
+                    error = IOError(errnoCode: result, reason: "connection reset (error set)")
```

Comment:
Rather than passing `reason` can you pass `function`? That'll give us better debug strings here.

Reply:
Just make one up. ;) The key advantage is that we get the textual description of the underlying errno, rather than your hard-coded guess.

Reply:
that's not how it's intended. The reason is there so it gives you a _better_ message. I acknowledge that we had a bug here where that was worse. In case where you don't want to provide a reason as it would be `"connect failed, errno \(errno)"`, you can use `function: #function` as that's shorter.